Module 5 docker build stages and added docker compose

jatin-28 opened the pull request:
Added support for running in a test container, but doesnt run the selenium tests

Commit:
Module 5: Created and launch todo app in docker

Commit:
Module 4: Added gunicorn to vagrant image and as a

Commit:
Module 4: Added docker ignore file and readme doc

Commit:
Module 5: Adding docker build stages

Commit:
Refactored todoapp and moved python code into its own source package

Commit:
Module 5: Refactored Docker file to contain production and development images. In addition, adding the foundation to support running tests

Commit:
Module 5: Added docker compose

hugh-emerson commented on .dockerignore:
```diff
@@ -0,0 +1,151 @@
+### Configuration files ###
+.env
+.env.template
+.env.test
```
You shouldn't ignore .env.test - the unit/integration tests are supposed to depend on it and there shouldn't be any secrets in there.

hugh-emerson commented on .dockerignore:
```diff
@@ -0,0 +1,151 @@
+# python
```
I realise you must have found this as a recommended .dockerignore file somewhere, but I think you should remove the irrelevant parts or put all of your additions in a clear section

hugh-emerson commented on runGunicorn.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+LOG_PATH=/var/log/gunicorn
+poetry run gunicorn --workers=2 --threads=4 --worker-class=gthread --bind 0.0.0.0:5000  --error-logfile $LOG_PATH/gunicorn_error.log --log-file $LOG_PATH/gunicorn.log 'todoapp.wsgi:create_app()'
```
Instead of having such a long command in an sh file, you might find it neater to put your config in a py file and just have e.g. `poetry run gunicorn --config=gunicorn_config.py 'todoapp.wsgi:create_app()` in your Dockerfile.

hugh-emerson commented on Dockerfile:
```diff
@@ -0,0 +1,77 @@
+RUN poetry config virtualenvs.create false && \
+    poetry install --no-dev
+
+RUN cd /app
```
This should be `WORKDIR /app` but actually you already did that so you can just delete this line

hugh-emerson commented on Dockerfile:
```diff
@@ -0,0 +1,77 @@
+COPY todoapp /app/todoapp
+
+RUN poetry config virtualenvs.create false && \
+    poetry install --no-dev
```
You should put those COPY commands after the poetry install, to speed up rebuilding the image:
- Currently you need to re-run poetry install after changing any app files
- If these COPY layers are later, then you'd only need to re-run poetry install when the toml files change.

hugh-emerson commented on README.md:
````diff
@@ -65,11 +65,48 @@ TODO_BOARD_ID
+Building and running tests 
+```
+$ docker build --target runtests --tag todo-app:runtests .
+$ docker run --env-file ./.env.test -p 5300:5000 todo-app:runtests
````
Your README is actually missing any instructions on running the tests normally.

hugh-emerson replied:
Why are you publishing a port when just running tests?

Also, your unit/integration tests should use .env.test but your selenium tests make a real board and so need values from the real .env

This means you will need to run them in a separate command. Group your tests into two folders so you can run:
`docker run --env-file .env todo-app:runtests tests`
`docker run todo-app:runtests tests`

hugh-emerson commented on Dockerfile:
```diff
@@ -0,0 +1,77 @@
+RUN poetry config virtualenvs.create false && \
+    poetry install
+
+CMD ["poetry", "run", "pytest", "--junit-xml", "test_results.xml"]
```
This should be an ENTRYPOINT rather than CMD, so that you can pass it different options to it later. See my other comment about running selenium tests separately.

hugh-emerson commented on README.md:
````diff
@@ -65,11 +65,48 @@ TODO_BOARD_ID
+
+Build the container
+```
+$ docker build --tag todo-app .
````
If you don't specify a target, this will build the last stage which is not what you want here.

hugh-emerson commented on todoapp/app.py:
```diff
@@ -1,9 +1,10 @@
-from trello_request import *
+import os
```
Moving this file means .env also needs to have `FLASK_APP=todoapp/app` but .env.template hasn't been updated.

hugh-emerson commented on tests/test_selenium_app.py:
```diff
@@ -2,13 +2,12 @@
-import app
-import common
-from trello_request import TrelloRequest
+from todoapp.trello_request import TrelloRequest
```
I get an import error when running it in docker without an `__init__.py` file in the todoapp folder.

hugh-emerson commented on tests/test_selenium_app.py:
```diff
@@ -93,7 +92,9 @@ def test_app():
 def driver():
-    with webdriver.Firefox() as driver:
+    options = webdriver.FirefoxOptions()
+    # TODO trialing running tests in docker. options.add_argument('--headless')
```
Yes you want to run it in headless mode.

hugh-emerson commented on Dockerfile:
```diff
@@ -0,0 +1,77 @@
+RUN chmod 755 runGunicorn.sh && \
+    mkdir -p /var/log/gunicorn
+
+ENV FLASK_APP=app \
+    FLASK_ENV=production
```
FLASK_APP=app is wrong but fortunately gets overridden when you run the container with `--env-file`. 

You should remove it here.

hugh-emerson reviewed (approved):
I needed to add an `__init__.py` file (see my other comment) but otherwise the app worked.  I've left some other suggestions.

You're close with running tests in the container, but that's not required until module 7. I've left some pointers to get you closer.

You should fix up the README and add the `__init__.py` file but merge when you're ready.

Commit:
Module 6:
Updated readme to include running locally with tests
Fixed and optimised Docker file
Clean up environment settings
Setup up Docker file to be ready to run with unit and integration testing

Commit:
Module 5:
Updated readme to include running locally with tests
Fixed and optimised Docker file
Clean up environment settings
Setup up Docker file to be ready to run with unit and integration testing
